frontend/src/metabase/query_builder/components/DatasetEditor/DatasetQueryEditor.jsx

```diff
@@ -93,9 +93,4 @@ function DatasetQueryEditor({
 
 DatasetQueryEditor.propTypes = propTypes;
 
-export default React.memo(
-  DatasetQueryEditor,
```

Comment:
If remove this we need to deal we the editor being always rendered another way. Looking at this code, could we conditionally render `DatasetQueryEditor` when `isEditingQuery` is `false`? https://github.com/metabase/metabase/blob/45f8791994d8f1af9d6e340cabbb529436707ea9/frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx#L466

Reply:
@ranquild re-mounting a notebook editor can be expensive if a query has a lot of clauses (see #12378), that's part of the reason why this `React.memo` was originally introduced

Reply:
### Measurements
I [measured](https://docs.google.com/spreadsheets/d/18HbARyTek1orJVx8jY79_j8GCWA4YTcP4-3OlF-kz64/edit#gid=0) the performance using [`React.Profiler`](https://react.dev/reference/react/Profiler), dev build, [this query](https://github.com/metabase/metabase/assets/6830683/86d717a7-19b0-4cad-a5fa-0391ae72360b) and viewport 1685 x 1335.

<details>
<summary>
Patch with code used to acquire these measurements
</summary>

Use `window.stats = {}` to clear measurements, use `window.printStats()` to log them.

```diff
diff --git a/frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx b/frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx
index d1abc185d6..660fdb2f8d 100644
--- a/frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx
+++ b/frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx
@@ -1,4 +1,4 @@
-import React, { useEffect, useCallback, useMemo, useState } from "react";
+import React, { Profiler, useEffect, useCallback, useMemo, useState } from "react";
 import PropTypes from "prop-types";
 import { connect } from "react-redux";
 import { t } from "ttag";
@@ -463,28 +463,46 @@ function DatasetEditor(props) {
       <Root>
         <MainContainer>
           <QueryEditorContainer isResizable={isEditingQuery}>
-            <DatasetQueryEditor
-              {...props}
-              isActive={isEditingQuery}
-              height={editorHeight}
-              viewHeight={height}
-              onResizeStop={handleResize}
-            />
+            <Profiler
+              id="DatasetQueryEditor"
+              onRender={(id, phase, actualDuration) => {
+                window.stats = window.stats || {};
+                window.stats[id] = window.stats[id] || [];
+                window.stats[id].push({ phase, actualDuration });
+              }}
+            >
+              <DatasetQueryEditor
+                {...props}
+                isActive={isEditingQuery}
+                height={editorHeight}
+                viewHeight={height}
+                onResizeStop={handleResize}
+              />
+            </Profiler>
           </QueryEditorContainer>
           <TableContainer isSidebarOpen={!!sidebar}>
             <DebouncedFrame className="flex-full" enabled>
-              <QueryVisualization
-                {...props}
-                className="spread"
-                noHeader
-                queryBuilderMode="dataset"
-                isShowingDetailsOnlyColumns={datasetEditorTab === "metadata"}
-                hasMetadataPopovers={false}
-                handleVisualizationClick={handleTableElementClick}
-                tableHeaderHeight={isEditingMetadata && TABLE_HEADER_HEIGHT}
-                renderTableHeaderWrapper={renderTableHeaderWrapper}
-                scrollToColumn={focusedFieldIndex + scrollToColumnModifier}
-              />
+              <Profiler
+                id="QueryVisualization"
+                onRender={(id, phase, actualDuration) => {
+                  window.stats = window.stats || {};
+                  window.stats[id] = window.stats[id] || [];
+                  window.stats[id].push({ phase, actualDuration });
+                }}
+              >
+                <QueryVisualization
+                  {...props}
+                  className="spread"
+                  noHeader
+                  queryBuilderMode="dataset"
+                  isShowingDetailsOnlyColumns={datasetEditorTab === "metadata"}
+                  hasMetadataPopovers={false}
+                  handleVisualizationClick={handleTableElementClick}
+                  tableHeaderHeight={isEditingMetadata && TABLE_HEADER_HEIGHT}
+                  renderTableHeaderWrapper={renderTableHeaderWrapper}
+                  scrollToColumn={focusedFieldIndex + scrollToColumnModifier}
+                />
+              </Profiler>
             </DebouncedFrame>
             <TabHintToastContainer
               isVisible={isEditingMetadata && isTabHintVisible && !result.error}
diff --git a/frontend/src/metabase/query_builder/components/view/View.jsx b/frontend/src/metabase/query_builder/components/view/View.jsx
index 83d4a3708f..f8f5f01d59 100644
--- a/frontend/src/metabase/query_builder/components/view/View.jsx
+++ b/frontend/src/metabase/query_builder/components/view/View.jsx
@@ -1,7 +1,6 @@
 /* eslint-disable react/prop-types */
-import React from "react";
+import React, { Profiler } from "react";
 import { Motion, spring } from "react-motion";
-import _ from "underscore";
 import { t } from "ttag";
 
 import ExplicitSize from "metabase/components/ExplicitSize";
@@ -37,6 +36,13 @@ import ViewSidebar from "./ViewSidebar";
 import NewQuestionView from "./View/NewQuestionView";
 import QueryViewNotebook from "./View/QueryViewNotebook";
 
+window.printStats = () => {
+  Object.entries(window.stats).forEach(([key, value]) => {
+    const sum = value.reduce((result, { actualDuration }) => result + actualDuration, 0);
+    console.log(key, sum);
+  });
+}
+
 import {
   BorderedViewTitleHeader,
   NativeQueryEditorContainer,
@@ -464,7 +470,16 @@ class View extends React.Component {
     if (question.isDataset() && queryBuilderMode === "dataset") {
       return (
         <>
-          <DatasetEditor {...this.props} />
+          <Profiler
+            id="DatasetEditor"
+            onRender={(id, phase, actualDuration) => {
+              window.stats = window.stats || {};
+              window.stats[id] = window.stats[id] || [];
+              window.stats[id].push({ phase, actualDuration });
+            }}
+          >
+            <DatasetEditor {...this.props} />
+          </Profiler>
           <QueryModals {...this.props} />
         </>
       );
```

</details>

### Results
[Google doc](https://docs.google.com/spreadsheets/d/18HbARyTek1orJVx8jY79_j8GCWA4YTcP4-3OlF-kz64/edit#gid=0)

![image](https://github.com/metabase/metabase/assets/6830683/b817c289-cdef-468b-b833-af7a6832e450)

### Conclusion

Rendering `DatasetQueryEditor` conditionally in `DatasetEditor` is the best option.

Reply:
> ### Conclusion
> Rendering `DatasetQueryEditor` conditionally in `DatasetEditor` is the best option.

This will however break the height animation.

I feel like any attempt to make it work with JS will affect the performance. 
The animation makes sense only when there's not much to render. Otherwise (when performance is visibly lower) users won't see it anyway.

Can we just remove the animation?

#### Current, with animation

https://github.com/metabase/metabase/assets/6830683/9454a570-ed86-4e6e-99ea-e884523f4470

#### Conditional rendering, no animation

https://github.com/metabase/metabase/assets/6830683/bbe2aacd-ee92-439c-b184-ee2e30e57b84

Reply:
Glad we can use conditional rendering here. Thanks for an excellent investigation, @kamilmielnik 🙏 

> This will however break the height animation.
>
> I feel like any attempt to make it work with JS will affect the performance.
The animation makes sense only when there's not much to render. Otherwise (when performance is visibly lower) users won't see it anyway.
> 
> Can we just remove the animation?

This is a tradeoff, but it makes sense to me. We had to disable the notebook slide-in/out animation for similar reasons.
Could you please double-check this with Maz and Bruno?

Reply:
> Could you please double-check this with Maz and Bruno?

The consensus is that performance is more important than the animation ([related Slack thread](https://metaboat.slack.com/archives/C057T1QTB3L/p1685540949163939)).